### PR TITLE
Optical model updates in flight continued: more WF measurement plotting

### PR DESCRIPTION
**Pull request by mperrin**
- Add wavefront trending plots and initial documentation for those. Merging in a lot of code that was developed in notebooks outside of webbpsf. 
- Misc updates to utility functions etc. More plot flexibility in gridded PSF library display. 
- Add docs page showing plots of wavefronts over time early in cycle 1

### CODE REVIEW

File: webbpsf/opds.py [hunk 1]
```diff
@@ -3207,8 +3206,76 @@ def coeffs_to_seg_state(coeffs):
...
+
+
+    coeffs = poppy.zernike.opd_expand_segments(opd, aperture=combined_mask,
+                                               basis=jw_ptt_basis, nterms=54,iterations=4)
+    fit = poppy.zernike.compose_opd_from_basis(coeffs, basis=jw_ptt_basis, npix=npix)
+
+    fit[~combined_mask]=np.nan
+    if plot:
+        fig, ax = plt.subplots(figsize=(16,4), nrows=1, ncols=1)
+        show_opd_image(np.hstack((opd, fit, opd-fit)),ax=ax )
```
Comment (marcio-melendez) [inline webbpsf/opds.py]: `show_opd_image` is not in OPDS, it needs to be initialized `webbpsf.trending.show_opd_image`.

File: webbpsf/opds.py [hunk 2]
```diff
@@ -3207,8 +3206,76 @@ def coeffs_to_seg_state(coeffs):
...
+        Coefficients per mode, in order corresponding to the webbpsf.opds.JWST_WAS_PTT_Basis class,
+        which is {piston, xtilt, ytilt} repeated per segments in order
+
+    """
+    npix = opd.shape[0]
+
+    ote_lom, jw_ptt_basis = _get_lom(npix)
+
+    if ote_lom.opd.shape[0] != npix:
+        ote_lom = webbpsf.opds.OTE_Linear_Model_WSS(npix=npix)
```
Comment (marcio-melendez) [inline webbpsf/opds.py]: Need to `import webbpsf` at the beginning of OPDS, see comment below.

File: webbpsf/webbpsf_core.py [hunk 3]
```diff
@@ -1529,7 +1535,7 @@ def load_wss_opd(self, filename, output_path = None, backout_si_wfe=True, verbos
                 axes[0,1].set_xlabel(f"RMS: {utils.rms(sensing_fp_si_wfe * 1e9, ote_pupil_mask):.2f} nm rms")
 
                 axes[0,2].imshow(opdhdu[0].data + sensing_fp_ote_wfe * ote_pupil_mask , vmin=-vm, vmax=vm, cmap=matplotlib.cm.RdBu_r)
```
Comment (marcio-melendez) [inline webbpsf/webbpsf_core.py]: All the `imshow`  in this block (if plot), should have `origin = 'lower'` so they match the orientation in the example in [https://webbpsf.readthedocs.io/en/latest/jwst_measured_opds.html]

File: webbpsf/opds.py [hunk 4]
```diff
@@ -3207,8 +3206,76 @@ def coeffs_to_seg_state(coeffs):
...
+@functools.lru_cache
+def _get_lom(npix):
+    """Initialize a few things that are computationally expensive so we cache for multiple reuses"""
+
+    ote_lom = OTE_Linear_Model_WSS(npix=npix)
+    jw_ptt_basis = JWST_WAS_PTT_Basis()
+    return ote_lom, jw_ptt_basis
+
+
+def decompose_opd_segment_PTT(opd, plot=False):
```
Comment (marcio-melendez) [inline webbpsf/opds.py]: I tried using this function (`decompose_opd_segment_PTT`) by expanding the example in https://webbpsf.readthedocs.io/en/latest/jwst_measured_opds.html
For this, after loading the OPD, I inspected and used it with the decomposed function. 
This is how the OPD looks, `plt.imshow(nrc.pupilopd[0].data)`: 
image
Then I applied the function as:
`webbpsf.opds.decompose_opd_segment_PTT(nrc.pupilopd[0].data, plot = True)`
and I get:
(array([[nan, nan, nan, ..., nan, nan, nan],
        [nan, nan, nan, ..., nan, nan, nan],
        [nan, nan, nan, ..., nan, nan, nan],
        ...,
        [nan, nan, nan, ..., nan, nan, nan],
        [nan, nan, nan, ..., nan, nan, nan],
        [nan, nan, nan, ..., nan, nan, nan]]),
 array([ 7.99506642e-09,  1.55340648e-08, -1.45810110e-08, -1.64441644e-08,
        -1.77063423e-09, -1.91410874e-08, -2.05439222e-08, -2.98547854e-09,
        -8.20687819e-09, -4.84058661e-09,  1.16624158e-08,  3.19651998e-08,
         1.52231422e-08,  1.86759131e-08, -1.61536156e-08,  2.21041071e-09,
         5.61439423e-08,  1.81762701e-08, -3.43559173e-09, -1.04949073e-08,
        -1.57846742e-08,  1.88472787e-09,  8.24705443e-09, -9.52822306e-09,
         1.90445152e-08,  2.29304796e-08, -2.46303600e-08, -1.12180320e-08,
        -7.22103361e-09, -1.77367429e-08,  2.62729490e-09,  3.92728519e-09,
         1.12543984e-08, -7.95697084e-09,  4.30708180e-09, -8.53238522e-09,
         5.46708572e-09, -5.92302208e-09,  1.57581015e-09,  5.59286666e-09,
        -5.79568266e-09, -3.81148890e-10,  8.19012707e-09,  1.14422039e-09,
         1.23863254e-08, -2.30900109e-09, -4.10657550e-09, -9.72345849e-09,
         2.31787969e-09,  6.65293229e-09,  9.50394333e-09, -1.10150228e-08,
        -5.89912643e-09, -1.47712434e-08]))

image

The values seems reasonable, at least in terms of range of expected corrections, but the figure seems out of scale. Maybe it needs a better/dynamic scaling range for display.

**Comment by marcio-melendez** [comment]
At this point we don't need a MAST token to download the OPDs, I just tested this by downloading OPDs without the `mast_wss_login` function. 
So, we can remove `mast_wss_login` and from: 

-  [`mast_retrieve_opd`](https://github.com/mperrin/webbpsf-st/blob/optical_model_updates_2/webbpsf/mast_wss.py#L60)
- [retrieve_mast_opd_table](https://github.com/mperrin/webbpsf-st/blob/optical_model_updates_2/webbpsf/mast_wss.py#L364)

File: webbpsf/opds.py (hunk 1 above)
Comment (mperrin) [reply webbpsf/opds.py]: 👍

File: webbpsf/opds.py (hunk 4 above)
Comment (mperrin) [reply webbpsf/opds.py]: I think the display is getting confused about units, yeah. Looks like the default scaling range is expecting microns but the array values are in meters

Comment (mperrin) [reply webbpsf/opds.py]: I made the default better, and also allowed the user to change the vmax if desired: 

Screen Shot 2022-09-14 at 5 38 44 PM

**Comment by mperrin** [comment]
@obi-wan76 this is revised and ready for re-review, please

**Review by marcio-melendez — APPROVED** [review]
Looks good, thanks!
image

**Comment by marcio-melendez** [comment]
@mperrin I just noticed that the pupil orientation from my test is different than yours, I am not sure where are you setting the origin of the figure, webbpsf.trending.show_opd_image? Your example has the entrance pupil orientation so I am not sure if this is a change that already exist and will get merge as part of other PR. This is just a very minor detail, mostly for consistency.

**Comment by mperrin** [comment]
I have the image origin set in my Matplotlib default so it always has the correct origin=lower, for all plots everywhere. But yes I can add it in explicitly to that function call too.  A minor detail but like you say good for consistency!

**Comment by mperrin** [comment]
See #582 now

**Pull request merged**